tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/RawTensor.java

```diff
@@ -84,33 +86,37 @@ public ByteDataBuffer data() {
    */
   @Override
   public String toString() {
-    return String.format("%s tensor with shape %s", dtype.toString(), shape);
+    return String.format("%s tensor with shape %s", typeInfo.dataType(), shape);
   }
 
   /**
    * Allocates a new tensor in native memory of the given type, shape and size.
    *
    * <p>The size of the tensor must be at least large enough to contain all scalars for the
-   * given type and shape, i.e. <code>size >= dtype.byteSize() * shape.size()</code>. More memory
+   * given type and shape, i.e. <code>size >= type.byteSize() * shape.size()</code>. More memory
    * can be allocated to store also metadata within the tensor itself, e.g. a lookup table
    * in a string tensor.
    *
-   * @param dtype data type
+   * @param type tensor type class
    * @param shape shape of the tensor
-   * @param size size of the tensor
+   * @param size size in bytes of the tensor, or -1 to compute the size from the shape
    * @return allocated tensor
    */
-  static RawTensor allocate(DataType<?> dtype, Shape shape, long size) {
-    // Minimum requirements for datatypes of variable length cannot be verified in a relevant way so
-    // we only validate them for fixed length datatypes
-    if (!dtype.isVariableLength() && shape.size() * dtype.byteSize() > size) {
+  static RawTensor allocate(Class<? extends TType> type, Shape shape, long size) {
+    TensorTypeInfo<?> typeInfo = TensorTypeRegistry.find(type);
+    long effectiveSize = size;
+    if (effectiveSize < 0) {
+      effectiveSize = shape.size() * typeInfo.byteSize();
```

Comment:
What happens if `typeInfo.byteSize()` returns -1?

Reply:
Actually this `if/else` is kind of reversed, I'll shuffle it a bit to make sure that won't be a problem.